homeassistant/helpers/entity.py

```diff
@@ -241,6 +241,22 @@ class EntityDescription:
     unit_of_measurement: str | None = None
 
 
+@dataclass(frozen=True, slots=True)
```

Comment:
I think we can skip freezing this since its only used internally

https://docs.python.org/3/library/dataclasses.html#frozen-instances

> There is a tiny performance penalty when using frozen=True: [__init__()](https://docs.python.org/3/reference/datamodel.html#object.__init__) cannot use simple assignment to initialize fields, and must use object.__setattr__().

Reply:
Yes, but I'm not sure we should use dataclasses at all.
Could you do some profiling maybe?

Reply:
Time for tuple: 0.029531209002016112
Time for dataclass: 0.38559437499498017

Reply:
micro benchmark

```python
from collections.abc import Mapping
from dataclasses import dataclass
import timeit
from typing import Any

empty_dict = {}


@dataclass(frozen=True, slots=True)
class CalculatedState:
    """Container with state and attributes.
    Returned by Entity._async_calculate_state.
    """

    state: str
    # The union of all attributes, after overriding with entity registry settings
    attributes: dict[str, Any]
    # Capability attributes returned by the capability_attributes property
    capability_attributes: Mapping[str, Any] | None
    # Attributes which may be overridden by the entity registry
    shadowed_attributes: Mapping[str, Any]


def test_dataclass():
    # mypy fails to narrow the type in the above if statement above
    return CalculatedState("state", empty_dict, None, empty_dict)


def test_with_tuple():
    return ("state", empty_dict, None, empty_dict)


dataclass_time = timeit.timeit(test_dataclass)
tuple_time = timeit.timeit(test_with_tuple)


print(f"Time for tuple: {tuple_time}")
print(f"Time for dataclass: {dataclass_time}")
```